File: src/main/java/de/clickism/clicksigns/block/HorizontalFacingBlockWithEntity.java
package de.clickism.clicksigns.block;

import net.minecraft.core.Direction;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.BaseEntityBlock;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import org.jetbrains.annotations.NotNull;

import static net.minecraft.world.level.block.state.properties.BlockStateProperties.HORIZONTAL_FACING;

/**
 * Utility class for blocks that are horizontally rotatable, and also have a block entity
 */
@SuppressWarnings("deprecation")
public abstract class HorizontalFacingBlockWithEntity extends BaseEntityBlock {

    protected HorizontalFacingBlockWithEntity(Properties properties) {
        super(properties);
        this.registerDefaultState(this.stateDefinition.any()
                .setValue(HORIZONTAL_FACING, Direction.NORTH));
    }

    @Override
    public @NotNull BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(
                HORIZONTAL_FACING,
                rotation.rotate(state.getValue(HORIZONTAL_FACING))
        );
    }

    @Override
    public @NotNull BlockState mirror(BlockState state, Mirror mirror) {
        return state.rotate(mirror.getRotation(state.getValue(HORIZONTAL_FACING)));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(HORIZONTAL_FACING);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext context) {
        return this.defaultBlockState()
                .setValue(HORIZONTAL_FACING, context.getHorizontalDirection());
    }
}

File: src/main/kotlin/de/clickism/clicksigns/block/HorizontalFacingBlockWithEntity.kt
package de.clickism.clicksigns.block

import net.minecraft.world.level.block.BaseEntityBlock
import net.minecraft.world.level.block.Mirror
import net.minecraft.world.level.block.Rotation
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties.HORIZONTAL_FACING

abstract class HorizontalFacingBlockWithEntity(properties: Properties) : BaseEntityBlock(properties) {
    override fun rotate(
        state: BlockState,
        rotation: Rotation
    ): BlockState? {
        return state.setValue(HORIZONTAL_FACING, rotation.rotate(state.getValue(HORIZONTAL_FACING)))
    }

    override fun mirror(state: BlockState, mirror: Mirror): BlockState? {
        return state.rotate(mirror.getRotation(state.getValue(HORIZONTAL_FACING)))
    }
}